opt_einsum/paths.py

```diff
@@ -971,14 +970,112 @@ def _dp_compare_size(
 
     s = s1 | s2
     i = _dp_calc_legs(g, all_tensors, s, inputs, i1_cut_i2_wo_output, i1_union_i2)
-    mem = helpers.compute_size_by_dict(i, size_dict)
+    mem = compute_size_by_dict(i, size_dict)
     cost = max(cost1, cost2, mem)
     if cost <= cost_cap:
         if s not in xn or cost < xn[s][1]:
             if memory_limit is None or mem <= memory_limit:
                 xn[s] = (i, cost, (contract1, contract2))
 
 
+def _dp_compare_write(
+    cost1,
+    cost2,
+    i1_union_i2,
+    size_dict,
+    cost_cap,
+    s1,
+    s2,
+    xn,
+    g,
+    all_tensors,
+    inputs,
+    i1_cut_i2_wo_output,
+    memory_limit,
+    cntrct1,
+    cntrct2,
+):
+    """Like ``_dp_compare_flops`` but sieves the potential contraction based
+    on the total size of memory created, rather than the number of
+    operations, and so calculates that first.
+    """
+    s = s1 | s2
+    i = _dp_calc_legs(g, all_tensors, s, inputs, i1_cut_i2_wo_output, i1_union_i2)
+    mem = compute_size_by_dict(i, size_dict)
+    cost = cost1 + cost2 + mem
+    if cost <= cost_cap:
+        if s not in xn or cost < xn[s][1]:
+            if memory_limit is None or mem <= memory_limit:
+                xn[s] = (i, cost, (cntrct1, cntrct2))
+
+
+DEFAULT_COMBO_FACTOR = 64
+
+
+def _dp_compare_combo(
+    cost1,
+    cost2,
+    i1_union_i2,
+    size_dict,
+    cost_cap,
+    s1,
+    s2,
+    xn,
+    g,
+    all_tensors,
+    inputs,
+    i1_cut_i2_wo_output,
+    memory_limit,
+    cntrct1,
+    cntrct2,
+    factor=DEFAULT_COMBO_FACTOR,
+    combine=sum,
+):
+    """Like ``_dp_compare_flops`` but sieves the potential contraction based
+    on some combination of both the flops and size,
+    """
+    s = s1 | s2
+    i = _dp_calc_legs(g, all_tensors, s, inputs, i1_cut_i2_wo_output, i1_union_i2)
+    mem = compute_size_by_dict(i, size_dict)
+    f = compute_size_by_dict(i1_union_i2, size_dict)
+    cost = cost1 + cost2 + combine((f, factor * mem))
+    if cost <= cost_cap:
+        if s not in xn or cost < xn[s][1]:
+            if memory_limit is None or mem <= memory_limit:
+                xn[s] = (i, cost, (cntrct1, cntrct2))
+
+
+minimize_finder = re.compile("(flops|size|write|combo|limit)-*(\d*)")
+
+
+@functools.lru_cache(128)
+def _parse_minimize(minimize):
+    """This works out what local scoring function to use for the dp algorithm
+    as well as a `naive_scale` to account for the memory_limit checks.
+    """
+    if minimize == "flops":
```

Comment:
Please change to `if/elif/else` patterns here. Functionality it doesn't matter, but helps with readability.